controllers/release/adapter.go

```diff
@@ -119,99 +115,51 @@ func (a *Adapter) EnsureFinalizerIsAdded() (reconciler.OperationResult, error) {
-			}
-
-			a.logger.Info("Created release PipelineRun",
-				"PipelineRun.Name", pipelineRun.Name, "PipelineRun.Namespace", pipelineRun.Namespace)
-		}
-
-		return reconciler.RequeueOnErrorOrContinue(a.registerReleaseStatusData(pipelineRun, releaseStrategy))
+	// The deployment has to complete if the environment field in the ReleasePlanAdmission is set
+	releasePlanAdmission, err := a.loader.GetActiveReleasePlanAdmissionFromRelease(a.ctx, a.client, a.release)
+	if err == nil && releasePlanAdmission.Spec.Environment != "" && !a.release.HasDeploymentFinished() {
+		return reconciler.ContinueProcessing()
 	}
 
-	return reconciler.ContinueProcessing()
+	patch := client.MergeFrom(a.release.DeepCopy())
+	a.release.MarkReleased()
+	return reconciler.RequeueOnErrorOrContinue(a.client.Status().Patch(a.ctx, a.release, patch))
 }
 
-// EnsureReleasePipelineStatusIsTracked is an operation that will ensure that the release PipelineRun status is tracked
-// in the Release being processed.
-func (a *Adapter) EnsureReleasePipelineStatusIsTracked() (reconciler.OperationResult, error) {
-	if !a.release.HasStarted() || a.release.IsDone() {
+// EnsureReleaseDeploymentIsTracked is an operation that will ensure that the SnapshotEnvironmentBinding
+// Deployment status is tracked in the Release being processed.
+func (a *Adapter) EnsureReleaseDeploymentIsTracked() (reconciler.OperationResult, error) {
+	if !a.release.IsDeploying() || a.release.HasDeploymentFinished() {
 		return reconciler.ContinueProcessing()
 	}
 
-	pipelineRun, err := a.loader.GetReleasePipelineRun(a.ctx, a.client, a.release)
+	// Search for an existing binding
+	binding, err := a.loader.GetSnapshotEnvironmentBindingFromReleaseStatus(a.ctx, a.client, a.release)
 	if err != nil {
 		return reconciler.RequeueWithError(err)
 	}
-	if pipelineRun != nil {
-		return reconciler.RequeueOnErrorOrContinue(a.registerReleasePipelineRunStatus(pipelineRun))
+
+	// Do nothing if the release does not own the binding
+	if binding.GetAnnotations()[libhandler.TypeAnnotation] != a.release.GetObjectKind().GroupVersionKind().GroupKind().String() ||
+		binding.GetAnnotations()[libhandler.NamespacedNameAnnotation] != fmt.Sprintf("%s/%s", a.release.GetNamespace(), a.release.GetName()) {
+		return reconciler.ContinueProcessing()
 	}
 
-	return reconciler.ContinueProcessing()
+	return reconciler.RequeueOnErrorOrContinue(a.registerDeploymentStatus(binding))
 }
 
-// EnsureSnapshotEnvironmentBindingExists is an operation that will ensure that a SnapshotEnvironmentBinding
+// EnsureReleaseIsDeployed is an operation that will ensure that a SnapshotEnvironmentBinding
 // associated to the Release being processed exists. Otherwise, it will create a new one.
-func (a *Adapter) EnsureSnapshotEnvironmentBindingExists() (reconciler.OperationResult, error) {
-	if !a.release.HasSucceeded() || a.release.IsDeployed() {
+func (a *Adapter) EnsureReleaseIsDeployed() (reconciler.OperationResult, error) {
```

Comment:
We are missing the code to set the deployment as outdated. Didn't we said that we would do it in a separate PR? I can do it now if you want.

Reply:
It can be a separate PR, but I didn't realize we had this bug now until reviewing this pr. If we already have the bug, keep the bug until after this pr then we can fix it. No more adding to this pr